src/basic/fs-util.c

```diff
@@ -1114,8 +1114,27 @@ int openat_report_new(int dirfd, const char *pathname, int flags, mode_t mode, b
                 if (errno != EEXIST)
                         return -errno;
 
-                /* Hmm, so now we got EEXIST? So it apparently exists now? If so, let's try to open again
-                 * without the two flags. But let's not spin forever, hence put a limit on things */
+                /* Hmm, so now we got EEXIST? This can indicate two things. First, if the path points to a
+                 * dangling symlink, the first openat() will fail with ENOENT because the symlink is resolved
+                 * and the second openat() will fail with EEXIST because symlinks are not followed when
+                 * O_CREAT|O_EXCL is specified. Let's check for this explicitly and fall back to opening with
+                 * just O_CREAT and assume we're the ones that created the file. */
+
+                struct stat st;
+                if (fstatat(dirfd, pathname, &st, AT_SYMLINK_NOFOLLOW) < 0)
+                        return -errno;
+
+                if (S_ISLNK(st.st_mode)) {
+                        fd = openat(dirfd, pathname, flags | O_CREAT, mode);
+                        if (fd < 0)
+                                return -errno;
+
+                        *ret_newly_created = true;
+                        return fd;
+                }
```

Comment:
hu? i am not following here? I am pretty sure we should *not* follow symlimks when creating a file here.

What is this about precisely?

I think the fact that on unix open/O_CREAT follows symlinks is a terrible idea, we should never indulge it.

Can you explain why you made this particular change? What's the usecase for this?

Reply:
I want to use `openat_report_new()` in `xopenat()`. Making that change caused a test failure because of what I describe in the commit message and comment above.

Christian even added a new `fcntl` `F_CREATED` flag for me to solve this problem.

https://lore.kernel.org/all/20240724-work-fcntl-v1-0-e8153a2f1991@kernel.org/T/

Reply:
> I want to use openat_report_new() in xopenat(). Making that change caused a test failure because of what I describe in the commit message and comment above.

well, i get the commit msg, but i am not grokking why this would be good behaviour this way. creating a file "through" a symlink is an abomination, not something we should support. we should really fail in that case.

isn't this a case where the test case should be adjusted?

Reply:
well, the way i see it creating files "through" a symlink is always a bug. hence i think we should catch cases like that.